src/tests/find.hpp

```diff
@@ -0,0 +1,81 @@
+// ============================== FIND TESTS =============================== //
+// Project:         The Experimental Bit Algorithms Library
+// Name:            find.hpp
+// Description:     Tests for find, find_if, find_if_not algorithms 
+//                  for bit iterators
+// Creator:         Vincent Reverdy
+// Contributor(s):  Collin Gress [2019]
+// License: BSD 3-Clause License
+// ========================================================================== //
+#ifndef _FIND_TESTS_HPP_INCLUDED
+#define _FIND_TESTS_HPP_INCLUDED
+// ========================================================================== //
+
+
+// =============================== PREAMBLE ================================= //
+// C++ standard library
+// Project sources
+#include "test_root.cc"
+#include "bit.hpp"
+// Third-party libraries
+// Miscellaneous
+// ========================================================================== //
+
+TEMPLATE_TEST_CASE("find: is correct for trivial single word cases", "[find]", 
+    unsigned short, unsigned int, unsigned long, unsigned long long) {
+
+    TestType t = 7;
+    bit_iterator<TestType*> first(&t, 0);
+    bit_iterator<TestType*> last(&t + 1, 0);
+
+    auto pos = bit::find(first, last, bit::bit0); 
+    auto pos_expected = std::find(first, last, bit::bit0);
+
+    REQUIRE(pos == pos_expected);
+}
+
+TEMPLATE_TEST_CASE("find: finding first set bit works for vectors", "[find]",
+    unsigned short, unsigned int, unsigned long, unsigned long long) {
+
+    using vec_t = std::vector<TestType>;
+    using vec_iter_t = typename vec_t::iterator;
+    
+    vec_t vec = {0, 0, 8, 0};
+
+    bit_iterator<vec_iter_t> first(vec.begin());
+    bit_iterator<vec_iter_t> last(vec.end());
+
+    auto pos = bit::find(first, last, bit::bit1);
+    auto pos_expected = std::find(first, last, bit::bit1);
+
+    REQUIRE(pos == pos_expected);
+
+    vec_t vec2 = {static_cast<TestType>(-1), static_cast<TestType>(-1), 3, 0};
+    bit_iterator<vec_iter_t> first2(vec2.begin());
+    bit_iterator<vec_iter_t> last2(vec2.end());
+
+    pos = bit::find(first2, last2, bit::bit0);
+    pos_expected = std::find(first2, last2, bit::bit0);
+
+    REQUIRE(pos == pos_expected);
+}
+
+TEMPLATE_TEST_CASE("find: is ok for virtual/misaligned words", "[find]",
```

Comment:
I think that until Vincent finishes his test framework, we should default to the `TEMPLATE_PRODUCT_TEST_CASE` that you introduced a few weeks ago. I would prefer it if we had somewhat more rigorous tests in place before his framework rolls out, but that's up to everyone. At the very least, we should be testing it on all possible iterator types.

Reply:
My thinking was to implement basic logic checks for now and revisit things after the testing tool is done. I guess it depends on the level of quality we want to enforce on master in the short term.